Mathlib/Algebra/Order/Hom/Monoid.lean

```diff
@@ -0,0 +1,743 @@
+a function via this coercion.
+
+There is no `OrderGroupHom` -- the idea is that `OrderMonoidHom` is used.
+The constructor for `OrderMonoidHom` needs a proof of `map_one` as well as `map_mul`; a separate
+constructor `OrderMonoidHom.mk'` will construct ordered group homs (i.e. ordered monoid homs
+between ordered groups) given only a proof that multiplication is preserved,
+
+Implicit `{}` brackets are often used instead of type class `[]` brackets. This is done when the
+instances can be inferred because they are implicit arguments to the type `OrderMonoidHom`. When
+they can be inferred from the type it is faster to use this method than to use type class inference.
+
+## Tags
+
+ordered monoid, ordered group, monoid with zero
+-/
+
+
+open Function
+
+variable {F α β γ δ : Type _}
+
+section AddMonoid
+
+/-- `α →+o β` is the type of monotone functions `α → β` that preserve the `OrderedAddCommMonoid`
+structure.
+
+`OrderAddMonoidHom` is also used for ordered group homomorphisms.
+
+When possible, instead of parametrizing results over `(f : α →+o β)`,
+you should parametrize over `(F : Type*) [OrderAddMonoidHomClass F α β] (f : F)`.
+
+When you extend this structure, make sure to extend `OrderAddMonoidHomClass`. -/
+structure OrderAddMonoidHom (α β : Type _) [Preorder α] [Preorder β] [AddZeroClass α]
+  [AddZeroClass β] extends α →+ β where
+  /-- An `OrderAddMonoidHom` is a monotone function. -/
+  monotone' : Monotone toFun
+#align order_add_monoid_hom OrderAddMonoidHom
+
+-- mathport name: «expr →+o »
+/-- Infix notation for `OrderAddMonoidHom`. -/
+infixr:25 " →+o " => OrderAddMonoidHom
+
+section
+
+/-- `OrderAddMonoidHomClass F α β` states that `F` is a type of ordered monoid homomorphisms.
+
+You should also extend this typeclass when you extend `OrderAddMonoidHom`. -/
+class OrderAddMonoidHomClass (F : Type _) (α β : outParam <| Type _) [Preorder α] [Preorder β]
+  [AddZeroClass α] [AddZeroClass β] extends AddMonoidHomClass F α β where
+  /-- An `OrderAddMonoidHom` is a monotone function. -/
+  monotone (f : F) : Monotone f
+#align order_add_monoid_hom_class OrderAddMonoidHomClass
+
+end
+
+-- Instances and lemmas are defined below through `@[to_additive]`.
+end AddMonoid
+
+section Monoid
+
+/-- `α →*o β` is the type of functions `α → β` that preserve the `OrderedCommMonoid` structure.
+
+`OrderMonoidHom` is also used for ordered group homomorphisms.
+
+When possible, instead of parametrizing results over `(f : α →*o β)`,
+you should parametrize over `(F : Type*) [OrderMonoidHomClass F α β] (f : F)`.
+
+When you extend this structure, make sure to extend `OrderMonoidHomClass`. -/
+@[to_additive]
+structure OrderMonoidHom (α β : Type _) [Preorder α] [Preorder β] [MulOneClass α]
+  [MulOneClass β] extends α →* β where
+  /-- An `OrderMonoidHom` is a monotone function. -/
+  monotone' : Monotone toFun
+#align order_monoid_hom OrderMonoidHom
```

Comment:
Could you please add the additive versions of the #aligns? This will later be done automatically, but for now needs to be done by hand.

Reply:
It does seem to be done automatically actually, when I try adding `#align order_add_monoid_hom OrderAddMonoidHom` after this line, I get an error saying it's been done already.

Reply:
OK I take that back, it's only done automatically some of the time. I'll check all of them to be sure.